pkg/cmd/roachtest/tests/decommissionbench.go

```diff
@@ -0,0 +1,317 @@
+
+	if !tc.admissionControl {
+		extraNameParts = append(extraNameParts, "no-admission")
+	}
+
+	extraName := strings.Join(extraNameParts, "/")
+
+	r.Add(registry.TestSpec{
+		Name: fmt.Sprintf("decommissionBench/nodes=%d/cpu=%d/warehouses=%d%s",
+			tc.nodes, tc.cpus, tc.warehouses, extraName),
+		Owner:             registry.OwnerKV,
+		Cluster:           r.MakeClusterSpec(tc.nodes+1, spec.CPU(tc.cpus)),
+		Timeout:           timeout,
+		NonReleaseBlocker: true,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runDecommissionBench(ctx, t, c, tc, timeout)
+		},
+	})
+}
+
+// decommBenchTicker is a simple struct for encapsulating function
+// closures to be called before and after an operation to be benchmarked.
+type decommBenchTicker struct {
+	pre  func()
+	post func()
+}
+
+// runDecommissionBench initializes a cluster with TPCC and attempts to
+// benchmark the decommissioning of a single node picked at random. The cluster
+// may or may not be running under load.
+func runDecommissionBench(
+	ctx context.Context,
+	t test.Test,
+	c cluster.Cluster,
+	tc decommissionBenchSpec,
+	testTimeout time.Duration,
+) {
+	// node1 is kept pinned (i.e. not decommissioned/restarted), and is the node
+	// through which we run decommissions. The last node is used for the workload.
+	pinnedNode := 1
+	workloadNode := tc.nodes + 1
+	crdbNodes := c.Range(pinnedNode, tc.nodes)
+	c.Put(ctx, t.Cockroach(), "./cockroach", c.All())
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload", c.Node(workloadNode))
+	for i := 1; i <= tc.nodes; i++ {
+		startOpts := option.DefaultStartOpts()
+		startOpts.RoachprodOpts.ExtraArgs = append(startOpts.RoachprodOpts.ExtraArgs, fmt.Sprintf("--attrs=node%d", i))
+		c.Start(ctx, t.L(), startOpts, install.MakeClusterSettings(), c.Node(i))
+	}
+
+	database := "tpcc"
+	importCmd := fmt.Sprintf(`./cockroach workload fixtures import tpcc --warehouses=%d`, tc.warehouses)
+	workloadCmd := fmt.Sprintf("./workload run tpcc --warehouses=%d --duration=%s --histograms=%s/stats.json --tolerate-errors {pgurl:1-%d}", tc.warehouses, testTimeout, t.PerfArtifactsDir(), tc.nodes)
+	t.Status(fmt.Sprintf("initializing cluster with %d warehouses", tc.warehouses))
+	c.Run(ctx, c.Node(pinnedNode), importCmd)
+
+	SetAdmissionControl(ctx, t, c, tc.admissionControl)
+	{
+		db := c.Conn(ctx, t.L(), pinnedNode)
+		defer db.Close()
+
+		for _, stmt := range []string{
+			fmt.Sprintf(`SET CLUSTER SETTING kv.snapshot_rebalance.max_rate='%dMiB'`, tc.snapshotRate),
+			fmt.Sprintf(`SET CLUSTER SETTING kv.snapshot_recovery.max_rate='%dMiB'`, tc.snapshotRate),
+		} {
+			t.Status(stmt)
+			_, err := db.ExecContext(ctx, stmt)
+			if err != nil {
+				t.Fatal(err)
+			}
+			t.L().Printf("run: %s\n", stmt)
+		}
+
+		// Wait for initial up-replication.
+		err := WaitFor3XReplication(ctx, t, db)
+		require.NoError(t, err)
+	}
+
+	workloadCtx, workloadCancel := context.WithCancel(ctx)
+	m := c.NewMonitor(workloadCtx, crdbNodes)
+
+	if tc.load {
+		m.Go(
```

Comment:
Do these workloads have a ramp time? Should they? If so, should we wait until that ramp period is done before initiating the decommission?